sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/With.scala

```diff
@@ -112,6 +112,22 @@ object With {
     With(replaced(commonExprRefs), commonExprDefs)
   }
 
+  /**
+   * Helper function to create a [[With]] statement when push down filter.
+   * @param expr original expression
+   * @param replaceMap Replaced attributes and common expressions
+   */
+  def apply(expr: Expression, replaceMap: Map[Attribute, Expression]): With = {
```

Comment:
if the attribute only appears once in the expression, we can just inline, doesn't need to be a common expression.

Reply:
This logic in `PushPredicateThroughNonJoin.rewriteByWith()`.